File: src/us/achromaticmetaphor/imcktg/MorseIMelody.java
package us.achromaticmetaphor.imcktg;

import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintStream;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class MorseIMelody extends ToneGenerator {

  private static final int defaultWordsPerMinute = 20;
  private static final int defaultOctave = IMelodyFormat.defaultOctave;
  private static final String defaultTone = "a";

  public static void main(String [] argV) throws IOException {
    (new MorseIMelody()).writeIMelody((OutputStream) System.out, argV[0]);
  }

  public void writeIMelody(OutputStream out, String s) throws IOException {
    PrintStream pout = new PrintStream(out);
    writeIMelody(pout, s);
    if (pout.checkError())
      throw new IOException();
  }

  public void writeIMelody(PrintStream out, String s) {
    writeIMelodyHeader(out, s);
    morseMelody(out, Morse.morse(s));
    writeIMelodyFooter(out);
  }

  private void writeIMelodyHeader(PrintStream out, String s) {
    IMelodyFormat.writeRequiredHeaders(out);
    IMelodyFormat.writeNameHeaderMangled(out, s);
    IMelodyFormat.writeBeatHeader(out, beat);
    IMelodyFormat.writeStyleHeader(out, IMelodyFormat.styleContinuous);
    IMelodyFormat.writeVolumeHeader(out, IMelodyFormat.volumeMax);
  }

  private static void writeIMelodyFooter(PrintStream out) {
    IMelodyFormat.writeRequiredFooters(out);
  }

  private final Map<Character, String> tones;
  private final int beat;

  public MorseIMelody() {
    this(defaultWordsPerMinute);
  }

  public MorseIMelody(int wpm) {
    this(defaultOctave, defaultTone, wpm);
  }

  public MorseIMelody(int octave, String tone, int wpm) {
    if (wpm < 1 || wpm > 144)
      throw new IllegalArgumentException("invalid wpm : " + wpm);

    final int duration = wpm <= 36 ? 3 : wpm <= 72 ? 4 : 5;
    // The format defines BEAT as beats per minute, at common (4/4) time.
    beat = wpm * Morse.unitsPerWord * 4 / (1 << duration);

    final String note = IMelodyFormat.note(octave, tone);
    final String rest = IMelodyFormat.rest;
    final String dot_duration = IMelodyFormat.duration(duration);
    final String dash_duration = IMelodyFormat.durationDotted(duration - 1);
    final String pause_duration = dot_duration;
    final String double_pause_duration = IMelodyFormat.duration(duration - 1);

    final String dot = note + dot_duration;
    final String dash = note + dash_duration;
    final String pause = rest + pause_duration;
    final String double_pause = rest + double_pause_duration;

    Map<Character, String> m = new HashMap<Character, String>();
    m.put(Morse.dotChar, dot + pause);
    m.put(Morse.dashChar, dash + pause);
    m.put(Morse.pauseChar, double_pause);
    tones = Collections.unmodifiableMap(m);
  }

  private void morseMelody(PrintStream out, Iterable<String> mcs) {
    IMelodyFormat.writeBeginMelody(out);
    for (String s : mcs)
      morseMelody(out, s);
    IMelodyFormat.writeEndMelody(out);
  }

  private void morseMelody(PrintStream out, String s) {
    for (int i = 0; i < s.length(); i++)
      out.print(tones.get(s.charAt(i)));
    out.print(tones.get(Morse.pauseChar));
    out.print(IMelodyFormat.lineContinuation);
  }

  @Override
  public void writeTone(OutputStream out, String s, boolean extend) throws IOException {
    writeIMelody(out, extend ? s + Tone.morsePostPause : s);
  }

  @Override
  public String filenameExt() {
    return ".imy";
  }

  @Override
  public String filenameTypePrefix() {
    return "iMelody:" + beat + ":" + tones.get(Morse.dotChar) + ":";
  }

}

File: src/main/java/us/achromaticmetaphor/imcktg/MorseIMelody.kt
package us.achromaticmetaphor.imcktg

import java.io.IOException
import java.io.OutputStream
import java.io.PrintStream

private const val defaultWordsPerMinute = 20
private const val defaultTone = "a"

class MorseIMelody(octave: Int, tone: String, wpm: Int, private val repeatCount: Int) : ToneGenerator() {
    @Throws(IOException::class)
    fun writeIMelody(out: OutputStream, s: String) {
        val pout = PrintStream(out)
        writeIMelody(pout, s)
        if (pout.checkError()) throw IOException()
    }

    private fun writeIMelody(out: PrintStream, s: String) {
        writeIMelodyHeader(out, s)
        morseMelody(out, Morse.morse(s))
        writeIMelodyFooter(out)
    }

    private fun writeIMelodyHeader(out: PrintStream, s: String) {
        IMelodyFormat.apply {
            writeRequiredHeaders(out)
            writeNameHeaderMangled(out, s)
            writeBeatHeader(out, beat)
            writeStyleHeader(out, styleContinuous)
            writeVolumeHeader(out, volumeMax)
        }
    }

    private val tones: Map<Char, String>
    private val beat: Int

    constructor(repeatCount: Int) : this(defaultWordsPerMinute, repeatCount) {}
    constructor(wpm: Int, repeatCount: Int) : this(IMelodyFormat.defaultOctave, defaultTone, wpm, repeatCount) {}

    private fun morseMelody(out: PrintStream, mcs: Iterable<String>) {
        IMelodyFormat.apply {
            writeBeginMelody(out)
            if (repeatCount > 0) beginRepeatBlock(out)
            for (s in mcs) {
                s.map{ e -> tones[e] }.forEach{ t -> out.print(t) }
                out.print(tones[Morse.pauseChar])
                out.print(lineContinuation)
            }
            if (repeatCount > 0) endRepeatBlock(out, repeatCount)
            writeEndMelody(out)
        }
    }

    @Throws(IOException::class)
    override fun writeTone(out: OutputStream, s: String) = writeIMelody(out, s)
    override fun filenameExt() = ".imy"
    override fun filenameTypePrefix() = "iMelody:" + beat + ":" + tones[Morse.dotChar] + ":" + repeatCount + ":"

    companion object {
        @Throws(IOException::class)
        @JvmStatic
        fun main(argV: Array<String>) = MorseIMelody(0).writeIMelody(System.out as OutputStream, argV[0])
        private fun writeIMelodyFooter(out: PrintStream) = IMelodyFormat.writeRequiredFooters(out)
    }

    init {
        require(!(wpm < 1 || wpm > 144)) { "invalid wpm : $wpm" }
        require(repeatCount >= 0) { "invalid repeat count : $repeatCount" }
        val duration = if (wpm <= 36) 3 else if (wpm <= 72) 4 else 5
        // The format defines BEAT as beats per minute, at common (4/4) time.
        beat = wpm * Morse.unitsPerWord * 4 / (1 shl duration)
        IMelodyFormat.apply {
            tones = mapOf(
                    Morse.dotChar to note(octave, tone) + duration(duration) + rest + duration(duration),
                    Morse.dashChar to note(octave, tone) + durationDotted(duration - 1) + rest + duration(duration),
                    Morse.pauseChar to rest + duration(duration - 1)
            )
        }
    }
}